lib/astc_encode.cpp

```diff
@@ -0,0 +1,919 @@
+ *                              The texture image format's component size is not
+                                8-bits.
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture's images are 1D. Only 2D images can
+ *                              be supercompressed.
+ * @exception KTX_INVALID_OPERATION
+ *                              ASTC  compressor failed to compress image for any
+                                reason.
+ * @exception KTX_OUT_OF_MEMORY Not enough memory to carry out compression.
+ */
+extern "C" KTX_error_code
+ktxTexture_CompressAstcEx(ktxTexture* _This, ktxAstcParams* params) {
+    // FIXME: At the moment defaults to ktx2 textures only
+    assert(_This->classId == ktxTexture2_c && "Only support ktx2 ASTC.");
+    ktxTexture2* This = (ktxTexture2*)_This;
+
+    KTX_error_code result;
+
+    if (!params)
+        return KTX_INVALID_VALUE;
+
+    if (params->structSize != sizeof(struct ktxAstcParams))
+        return KTX_INVALID_VALUE;
+
+    if (This->supercompressionScheme != KTX_SS_NONE)
+        return KTX_INVALID_OPERATION; // Can't apply multiple schemes.
+
+    if (This->isCompressed)
+        return KTX_INVALID_OPERATION;  // Only non-block compressed formats
+                                       // can be encoded into an ASTC format.
+
+    if (This->_protected->_formatSize.flags & KTX_FORMAT_SIZE_PACKED_BIT)
+        return KTX_INVALID_OPERATION;
+
+    // Basic descriptor block begins after the total size field.
+    const uint32_t* BDB = This->pDfd+1;
+
+    uint32_t num_components, component_size;
+    getDFDComponentInfoUnpacked(This->pDfd, &num_components, &component_size);
+
+    if (component_size != 1)
+        return KTX_INVALID_OPERATION; // Can only deal with 8-bit components
+
+    if (This->pData == NULL) {
+        result = ktxTexture2_LoadImageData((ktxTexture2*)This, nullptr, 0);
+
+        if (result != KTX_SUCCESS)
+            return result;
+    }
+
+    ktx_uint32_t threadCount = params->threadCount;
+    if (threadCount < 1)
+        threadCount = 1;
+
+    ktx_uint32_t transfer = KHR_DFDVAL(BDB, TRANSFER);
+    bool sRGB = transfer == KHR_DF_TRANSFER_SRGB;
+
+    VkFormat vkFormat = astcVkFormat(params->blockDimension, sRGB);
+
+    // Create a prototype texture to use for calculating sizes in the target
+    // format and, as useful side effects, provide us with a properly sized
+    // data allocation and the DFD for the target format.
+    ktxTextureCreateInfo createInfo;
+    createInfo.glInternalformat = 0;
+    createInfo.vkFormat = vkFormat;
+    createInfo.baseWidth = This->baseWidth;
+    createInfo.baseHeight = This->baseHeight;
+    createInfo.baseDepth = This->baseDepth;
+    createInfo.generateMipmaps = This->generateMipmaps;
+    createInfo.isArray = This->isArray;
+    createInfo.numDimensions = This->numDimensions;
+    createInfo.numFaces = This->numFaces;
+    createInfo.numLayers = This->numLayers;
+    createInfo.numLevels = This->numLevels;
```

Comment:
You need

```
This->numLevels = max(1, This->numLevels)
```

before this so avoid numLevels = 0 in the final texture. You can remove line 743 then. Basically moving it here.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/8374a6988565a9c4ba280d667de5c305d3c67cb1